File: python/2015/04/solution.py
import hashlib

data = b'iwrupvqb'

base_hash = hashlib.md5(data)

found5 = False
found6 = False
five = 0
six = 0

i = 0

while True:
    i += 1
    i_str = bytes(str(i), 'utf8')
    cp = base_hash.copy()
    cp.update(i_str)
    hx = cp.digest()

    if not found5 and hx[0] == 0 and hx[1] == 0 and hx[2]&0xF0 == 0:
        five = i
        found5 = True
    if not found6 and hx[0] == 0 and hx[1] == 0 and hx[2] == 0:
        six = i
        found6 = True

    if found5 and found6:
        break

print(five)
print(six)

File: go/2015/04/solution.go
package main

import (
	"aoc/utils"
	"crypto/md5"
	"fmt"
	"strconv"
)

func main() {
	data := utils.Read()
	found5 := false
	found6 := false
	five := 0
	six := 0

	i := 0
	for {
		i++
		hash := md5.New()
		hash.Write([]byte(data))
		hash.Write([]byte(strconv.Itoa(i)))
		sum := hash.Sum(nil)

		if !found5 && sum[0] == 0 && sum[1] == 0 && sum[2]&0xF0 == 0 {
			found5 = true
			five = i
		}

		if !found6 && sum[0] == 0 && sum[1] == 0 && sum[2] == 0 {
			found6 = true
			six = i
		}

		if found5 && found6 {
			break
		}
	}

	fmt.Println(five)
	fmt.Println(six)
}
